### DISCUSSION TITLE
Text selection and word boundaries

### DISCUSSION THREAD

**Discussion by HeyMyian** [vkbo/novelWriter#1485] [Ideas]
Consider following text in the novelWriter editor or viewer:

`He knew—somehow.`

The dash is an em dash `—`, and no spaces around it is the American English way of typing this. If I want to select the word "somehow" with double-click or keyboard or context menu, it selects both words around the em dash, because as I understand the em dash is not recognized as a word boundary. Notice how in this GitHub editor, selecting the single word works.

Other symbols that are not recognized as word boundaries: Underscores, ellipses, double quotation marks (what would the correct behavior be for British English with single quotation marks that are the same symbols as apostrophes?).

It's a super minor issue, but it's a bit like death by a thousand cuts. Is there any way to have novelWriter recognize these symbols as word boundaries? Is that even a sensible question to ask, or is this some sort of standard behavior?

Related to #1415 maybe?

**Comment by vkbo** [comment]
No, it's not related to #1415 at all. The spell checker is something I control, so I could easily fix that. The central functionality of the QTextEdit, I have little control over, which is the issue in this case. This is not an issue I can fix unfortunately.

Using a dash with no spaces is an American style, and I suspect that since Qt is European made, they may not have considered this, or chosen to ignore it for simplicity. In either case, this would have to be reported to Qt. It may even have been fixed in Qt6, which novelWriter cannot use yet.

**Reply by HeyMyian** [reply]
Thanks for the info. I dug a bit and there appears to be a report for the issue for Qt6 from March 2022, meaning it's not yet fixed: https://bugreports.qt.io/browse/QTBUG-101372

**Reply by vkbo** [reply]
Thanks for the link.

Now that I'm thinking about it, I may be able to capture the double-click and direct it to my own selector function. The problem is that I'd need to capture all the possible ways to select a word, like `Ctrl+Shift+Left/Right` and `Ctrl+BackSpace` etc.

Hopefully they'll just fix the underlying word selector mechanism in the QTextCursor. I also use it for the auto-select feature for applying formatting.

**Reply by HeyMyian** [reply]
> my own selector function

If this is happening, I would like to throw in the idea that upon selecting a word, identical words in the text will be highlighted too (with a slightly different shade than the select color). You probably know this feature from coding IDEs like VSCode. I think it would be rather neat to have for creative writing too, to easily spot word repetitions. I don't know if it's at all possible.

**Reply by vkbo** [reply]
It would be possible to do this with the syntax highlighter, but currently it also runs the spell checker, which is slow, and for some reason running the rehighlight command is extremely slow when called from Python. It's a "bug" I've struggled with a lot (not sure if it's a bug or just an issue with the Python/C++ interface).

While this is a great idea, it isn't doable in practice. This is also why the search feature doesn't highlight all results right away. I wanted to do it there too.

When I have time to work on my C++ sister app to novelWriter, I'll certainly look into implementing it there.

**Reply by HeyMyian** [reply]
I'm really getting to understand why you want to rewrite novelWriter with a different language :smile:

**Reply by vkbo** [reply]
Yeah. the frustration with not being able to enhance the text editor is a major reason. The new project is a rich text editor, but otherwise it is very similar to novelWriter. I'm able to implement a bunch of features that have been requested here, but aren't really feasible.

The problem with C++ is that it is a lot more work to code, because Python is a much easier and less verbose language, and I know Python a hell of a lot better than C++. Python is one of the main languages I use professionally.
